flashlight/fl/tensor/backend/jit/ShapeInference.cpp

```diff
@@ -0,0 +1,213 @@
+    }
+  } else {
+    std::ostringstream oss;
+    oss << "[JitBackend::transpose] Invalid axes: " << axes
+        << " for shape: " << inputShape;
+    throw std::runtime_error(oss.str());
+  }
+  return Shape(outputDims);
+}
+
+Shape inferTileOutputShape(const Shape& inputShape, const Shape& tileDims) {
+  std::vector<Dim> paddedTensorDims = inputShape.get();
+  std::vector<Dim> paddedTileDims = tileDims.get();
+  const auto inputRank = inputShape.ndim();
+  const auto tileRank = tileDims.ndim();
+  if (inputRank > tileRank) {
+    const auto diff = inputRank - tileRank;
+    paddedTileDims.insert(paddedTileDims.end(), diff, 1);
+  } else {
+    const auto diff = tileRank - inputRank;
+    paddedTensorDims.insert(paddedTensorDims.end(), diff, 1);
+  }
+  std::vector<Dim> outputDims;
+  for (unsigned i = 0; i < paddedTensorDims.size(); i++) {
+    outputDims.push_back(paddedTensorDims[i] * paddedTileDims[i]);
+  }
+  return Shape(outputDims);
+}
+
+Shape inferConcatenateOutputShape(
+    const std::vector<Tensor>& tensors,
+    const unsigned axisToConcat) {
+  // TODO need a nice way to construct empty tensor for wrapped backend
+  if (tensors.empty()) {
+    throw std::runtime_error(
+        "[inferConcatenateOutputShape] Nothing to concatenate");
+  }
+  const auto shapes = broadcastShapesToMaxRank(tensors, axisToConcat + 1);
+  const unsigned outputRank = shapes.front().ndim();
+  if (axisToConcat >= outputRank) {
+    std::ostringstream oss;
+    oss << "[inferConcatenateOutputShape] Axis too big: " << axisToConcat;
+    throw std::runtime_error(oss.str());
+  }
+  std::vector<Dim> outputDims;
+  for (unsigned axis = 0; axis < outputRank; axis++) {
+    Dim concatenatedDim = 0;
+    if (axis == axisToConcat) {
+      for (const auto& shape : shapes) {
+        concatenatedDim += shape.dim(axis);
+      }
+    } else {
+      concatenatedDim = shapes.front().dim(axis);
+    }
+    outputDims.push_back(concatenatedDim);
+  }
+  return Shape(outputDims);
+}
+
+Shape inferPadOutputShape(
+    const Shape& inputShape,
+    const std::vector<std::pair<int, int>>& padWidths) {
+  std::vector<Dim> outputDims = inputShape.get();
+  if (padWidths.size() > static_cast<size_t>(inputShape.ndim())) {
+    throw std::runtime_error("[inferPadOutputShape] too many paddings");
+  }
+  for (unsigned axis = 0; axis < padWidths.size(); axis++) {
+    const auto& [beforeDim, afterDim] = padWidths[axis];
+    outputDims[axis] += beforeDim + afterDim;
+  }
+  return Shape(outputDims);
+}
+
+Shape inferMatmulOutputShape(
+    const Shape& lhsShape,
+    const Shape& rhsShape,
+    MatrixProperty lhsProp,
+    MatrixProperty rhsProp) {
+  std::vector<Dim> lhsDims = lhsShape.get();
+  std::vector<Dim> rhsDims = rhsShape.get();
+  const auto lhsRank = lhsDims.size();
+  const auto rhsRank = rhsDims.size();
+  const bool isLhsScalarOrVector = lhsRank <= 1;
+  const bool isRhsScalarOrVector = rhsRank <= 1;
+  if (isLhsScalarOrVector) { // pad to (1 x 1/K)
```

Comment:
Why does this help with oneDNN? Fine with it, but worth noting that there are many cases in which frameworks will do this broadcast implicitly, and where it might actually be useful to distinguish between a scalar (which is typically 0D shape-wise) and a 1-dimensional tensor.